### ISSUE TITLE
Keeper crashes after migration from Zookeeper

### ISSUE THREAD

**Issue by paoloxide** [ClickHouse/ClickHouse#43872]
**Describe the issue**
I created a setup of clickhouse server nodes connected to zookeeper nodes and a set of clickhouse-keeper nodes where I will be migrating to. Upon following this documentation: https://clickhouse.com/docs/en/operations/clickhouse-keeper/

<img width="1220" alt="image" src="https://user-images.githubusercontent.com/6601604/205260511-c5204ab8-ad0e-4444-84d3-aa5068f8b7db.png">

I was able to successfully (based from the migration logs) migrate the zookeeper snapshot to clickhouse keeper snapshot, I updated the config of the Clickhouse server (while it is turned off) to connect to Clickhouse keeper nodes (port 9181). After turning it on, I was able to view the replica table values but noticed that the table became read-only:

┌─table──┬─is_readonly─┐
  │ table1          │                           1 │
└── ───┴─── ────┘

With this, I cannot insert anymore to the table1 because it became readonly.

**Comment by filimonov** [comment]
```
SELECT zookeeper_exception FROM system.replicas;

SELECT * FROM system.zookeeper WHERE path=(SELECT zookeeper_path FROM system.replicas WHERE database = '' and table = '...')
```

**Comment by paoloxide** [comment]
Hi @filimonov, tried running the query statements you sent, but the first one gave me an empty zookeeper_exception column. Therefore, the second statement is not possible and returns an error.

**Comment by tavplubix** [comment]
> the second statement is not possible and returns an error.

What error exactly?

**Comment by paoloxide** [comment]
Hi @tavplubix, I am the one from the telegram chat who sent this. Thank you for updating the ticket.

**Comment by antonio2368** [comment]
@paoloxide can you provide please full logs after starting an instance with trace level enabled?
Including `raft_logs_level` (https://clickhouse.com/docs/en/operations/clickhouse-keeper/#configuration)

**Comment by paoloxide** [comment]
@filimonov I was able to re-simulate the scenario I originally posted and I am seeing as a result of the query you mentioned:

<img width="1247" alt="image" src="https://user-images.githubusercontent.com/6601604/205887891-1fc26ac4-7ae3-4ce4-a6e8-36f8a8509aad.png">

**Comment by paoloxide** [comment]
@antonio2368 Please see attached full logs after restarting the three keeper nodes (after conversion via clickhouse keeper converter)
[mchkeeper2-err.log](https://github.com/ClickHouse/ClickHouse/files/10165368/mchkeeper2-err.log)

**Comment by antonio2368** [comment]
can you provide regular logs also somehow? (not the errors)
I see a lot of networking issues. Did you set everything correctly from the networking side? Hostnames can't be found for some reason.

**Comment by paoloxide** [comment]
Hi @antonio2368, okay attaching it now. Yes, the networking issue is more likely when nodes are restarted all together, but after they get restarted, I get 'imok' from the running keepers. Here is the full log:

[mchkeeper2.log](https://github.com/ClickHouse/ClickHouse/files/10165439/mchkeeper2.log)

**Comment by antonio2368** [comment]
I suspect something strange happened with logs because of the steps you took.
Based on the logs it looks like you started CH server with empty embedded Keepers, they produced some logs (because of ZK client heartbeats) and then you moved the generated snapshots to a Keeper node while the previous Keeper is running. This created an inconsistent RAFT state producing the errors you have.
Try stopping all CH instances, clean logs and snapshots from the coordination folders of each Keeper, move the generated ZK snapshot into snapshot folder inside coordination folder or generate it with `clickhouse-keeper-converter` program, and then start all CH servers.

**Comment by paoloxide** [comment]
Hi @antonio2368, here is the current setup for further context so maybe you can help pinpoint at what step I might be doing wrong:

Goal setup:
Migrate from 3 Zookeeper Nodes to 3 Clickhouse Keeper Nodes
4 Clickhouse Nodes connected to an external 3 Clickhouse Keeper Nodes

Steps I took:
1. Started 3 Zookeeper nodes

2. Started 4 Clickhouse nodes (without keeper config, connected to 3 Zookeeper nodes) - these clickhouse nodes will be used as the actual db for read/write

3. Started 3 Clickhouse nodes (with keeper config) - these nodes will only be used as an external clickhouse keeper and will not be utilized for db read/write

4. Created tables and inserted some values to the 4 clickhouse nodes.
5. Verify replicas. 
6. Stop 4 Clickhouse nodes that has data
7. Verify the leaders in Zookeeper and Clickhouse Keeper nodes.
8. Stop 3 Zookeeper nodes
9. Copy the /datalog and /data from Zookeeper nodes to the leader Clickhouse keeper node.

10. Execute the conversion in the leader Clickhouse keeper node:
clickhouse-keeper-converter --zookeeper-logs-dir /zoo2/datalog/version-2 --zookeeper-snapshots-dir /zoo2/data/version-2 --output-dir /var/lib/clickhouse/coordination/snapshots 

After conversion, it shows a log that looks like this:
Totally have 3 snapshots, will use latest
Deserializing storage snapshot /zoo2/data/version-2/snapshot.200000020
Magic deserialized, looks OK
Sessions and timeouts deserialized
ACLs deserialized
Deserializing data from snapshot
Finished, snapshot ZXID 8589934624
Totally have 2 logs
Deserializing log /zoo2/datalog/version-2/log.100000001
Header looks OK
Finished /zoo2/datalog/version-2/log.100000001 deserialization, totally read 143 records
Deserializing log /zoo2/datalog/version-2/log.300000001
Header looks OK
Finished /zoo2/datalog/version-2/log.300000001 deserialization, totally read 72 records
Snapshot serialized to path:/var/lib/clickhouse/coordination/snapshots/snapshot_12884901960.bin.zstd

After this, I tried restarting the Clickhouse Keeper nodes, and the error shows.

**Comment by antonio2368** [comment]
> Started 3 Clickhouse nodes (with keeper config) - these nodes will only be used as an external clickhouse keeper and will not be utilized for db read/write

They should be started only after you have generated snapshot on the leader.

Also, you should start Keeper as a standalone service or at least remove `zookeeper` configuration from them so they don't connect to their own Keeper cluster.

**Comment by paoloxide** [comment]
@antonio2368 tried the suggestions above but the error still persists after conversion from zookeeper. And if I try to connect them to the clickhouse server (even if they are not starting up properly), the table becomes readonly and the leader keeper dies and gets reassigned.

**Comment by antonio2368** [comment]
Did you clean the logs from the new Keeper nodes?

Steps should be:

Start 3 Zookeeper nodes

Start 4 Clickhouse nodes (without keeper config, connected to 3 Zookeeper nodes) - these clickhouse nodes will be used as the actual db for read/write

Created tables and inserted some values to the 4 clickhouse nodes.

Verify replicas.

Stop 4 Clickhouse nodes that has data

Verify the leaders in Zookeeper nodes.

Stop 3 Zookeeper nodes

Execute the conversion in the leader Clickhouse keeper node:
clickhouse-keeper-converter --zookeeper-logs-dir /zoo2/datalog/version-2 --zookeeper-snapshots-dir /zoo2/data/version-2 --output-dir /var/lib/clickhouse/coordination/snapshots

Start 3 Clickhouse nodes (with keeper config)

**Comment by paoloxide** [comment]
Hi @antonio2368, followed the steps but still getting the error after running the conversion command:

clickhouse-keeper-converter --zookeeper-logs-dir /zoo2/datalog/version-2 --zookeeper-snapshots-dir /zoo2/data/version-2 --output-dir /var/lib/clickhouse/coordination/snapshots

After conversion, I try to restart the 3 clickhouse-keeper nodes and the leader always dies with the same errors as before.

**Comment by antonio2368** [comment]
> I try to **restart** the 3 clickhouse-keeper nodes

Sorry for asking the same thing again, but they were stopped and with no data in coordination folder before running `clickhouse-keeper-converter`?

**Comment by paoloxide** [comment]
Hi @antonio2368, no worries. Yes, there are no data because the clickhouse-keeper nodes are created fresh and new before running the clickhouse-keeper-converter.

**Comment by antonio2368** [comment]
> nodes are created fresh and new before running the clickhouse-keeper-converte

What do you mean by created?

**Comment by paoloxide** [comment]
The clickhouse-keeper nodes are different machines from clickhouse-server nodes. They are created separately, 3 clickhouse-server nodes and 3 clickhouse-keeper nodes.

**Comment by paoloxide** [comment]
Thank you @antonio2368 for your wonderful insights.

I was able to implement the fix by implementing the following:

- Create separate instances of clickhouse keeper nodes which will perform the conversion.
   - Use a different path for conversion instead of /var/lib/clickhouse/coordination/snapshot, I used a /temp/keeper-snapshot
- Create another separate instances of clickhouse keeper nodes from a custom image built where the converted snapshot will be copied inside before running the instance.
- The image should also use a USER clickhouse or force chown the /var/lib/clickhouse/coordination directory and its subdirectories.
